TailLight/SetTaillightBlack.cpp

```diff
@@ -0,0 +1,268 @@
+            return status;
+        }
+
+        // We're warned not to process a symlink path name returned from 
+        // IoRegisterDeviceInterface so we need to use a unique device property
+        //  - the device object name - to see if we can send the IRP down with
+        //  minimal bus traffic.
+        status = TryToOpenIoTarget(hidTarget, symLink);
+    }
+
+    if (NT_SUCCESS(status)) {
+        
+        UNICODE_STRING theirPDOName = {};
+
+        // The PDO name will be attached to the IO target
+        // and thus deleted at the end of the function.
+        theirPDOName = GetTargetPropertyString(hidTarget,
+            DevicePropertyPhysicalDeviceObjectName);
+
+        if (theirPDOName.MaximumLength > 0) {
+            if (!RtlEqualUnicodeString(&pDeviceContext->PdoName,
+                &theirPDOName,
+                TRUE)) {
+                status = STATUS_NOT_FOUND;
+                goto ExitAndFree;
+            }
+        }
+
+        WDFREQUEST  request = NULL;
+
+        status = WdfRequestCreate(WDF_NO_OBJECT_ATTRIBUTES,
+            hidTarget,
+            &request);
+
+        if (!NT_SUCCESS(status)) {
+            KdPrint(("TailLight: WdfRequestCreate failed 0x%x\n", status));
+            goto ExitAndFree;
+        }
+
+        WdfRequestSetCompletionRoutine(
+            request,
+            SetBlackCompletionRoutine,
+            WDF_NO_CONTEXT);
+
+        TailLightReport  report = {};
+        report.Blue = 0x0;
+        report.Green = 0x0;
+        report.Red = 0x0;
+
+        // Set up a WDF memory object for the IOCTL request
+        WDF_OBJECT_ATTRIBUTES mem_attrib = {};
+        WDF_OBJECT_ATTRIBUTES_INIT(&mem_attrib);
+        mem_attrib.ParentObject = request; // auto-delete with request*/
+
+        WDFMEMORY InBuffer = 0;
+        BYTE* pInBuffer = nullptr;
+
+        status = WdfMemoryCreate(&mem_attrib,
+            NonPagedPoolNx,
+            POOL_TAG,
+            sizeof(TailLightReport),
+            &InBuffer,
+            (void**)&pInBuffer);
+
+        if (!NT_SUCCESS(status)) {
+            KdPrint(("TailLight: WdfMemoryCreate failed: 0x%x\n", status));
+            goto ExitAndFree;
+        }
+
+        // TODO: Wondering if we just cant cast pInBuffer as a TailLightReport
+        RtlCopyMemory(pInBuffer, &report, sizeof(TailLightReport));
+
+        // Format the request as write operation
+        status = WdfIoTargetFormatRequestForIoctl(hidTarget,
+            request,
+            IOCTL_HID_SET_FEATURE,
+            InBuffer,
+            NULL,
+            0,
+            0);
+
+        if (!NT_SUCCESS(status)) {
+            KdPrint(("TailLight: WdfIoTargetFormatRequestForIoctl failed: 0x%x\n", status));
+            goto ExitAndFree;
+        }
+
+        if (!WdfRequestSend(request, hidTarget, WDF_NO_SEND_OPTIONS)) {
+            WdfObjectDelete(request);
+            request = NULL;
+            status = STATUS_UNSUCCESSFUL;
+        }
+    }
+
+ExitAndFree:
```

Comment:
Please use the `WDFIOTARGET_Wrap` RAII wrapper for `hidTarget` so that you avoid goto in this function. I´ve just moved this wrapper to vfeature.h to ease usage.